[WIP] Add a pymatgen.io.openff module

orionarcher opened the pull request:
## Summary

This PR adds basic integration with the `openff.toolkit` Molecule representation. It is intended to be use in the `classical_md` workflows in `atomate2` and `emmet`.

Currently, this functionality is downstream in a PR to [atomate2](https://github.com/materialsproject/atomate2/pull/782). As basic IO functionality, I would like to migrate it upstream to `pymatgen`. Especially because some functions will also be used in the `classical_md` `emmet` `Builder`.

Major changes:

- Several utility functions that support IO between `pymatgen.Molecule`/`pymatgen.MoleculeGraph` <-> `openff.toolkit.Molecule`

## Todos

`openff.toolkit` is only installable via conda. I am not sure the best way to add it to CI. I'd appreciate guidance from the core developers on this.

## Checklist

- [ ] Google format doc strings added. Check with `ruff`.
- [x] Type annotations included. Check with `mypy`.
- [x] Tests added for new features/fixes.
- [x] If applicable, new classes/functions/modules have [`duecredit`](https://github.com/duecredit/duecredit) `@due.dcite` decorators to reference relevant papers by DOI ([example](https://github.com/materialsproject/pymatgen/blob/91dbe6ee9ed01d781a9388bf147648e20c6d58e0/pymatgen/core/lattice.py#L1168-L1172))

Tip: Install `pre-commit` hooks to auto-check types and linting before every commit:

```sh
pip install -U pre-commit
pre-commit install
```

janosh reviewed (commented):
cool, excited to see this! 👍 

is the issue here that [`openff` is only installable](https://docs.openforcefield.org/projects/toolkit/en/stable/installation.html#installing-from-source) with `conda`/`mamba`?
we could try installing from source in CI

orionarcher commented:
> we could try installing from source in CI

Sounds good! I'll try adding that to the CI.

orionarcher commented:
OpenBabel is a core dependency of the OpenFF toolkit, and I don't believe it is pip installable. I believe that's the reason openff doesn't support pip). I believe we'd either have to build it from source OR use conda to install it. 

I know this is a problem in other parts of the MP ecosystem and I am curious how you'd recommend handling it?

DanielYang59 commented:
> OpenBabel is a core dependency of the OpenFF toolkit, and I don't believe it is pip installable. 

I'm interested in hearing about this as well, as `OpenBabel` is not installed in CI as well https://github.com/materialsproject/pymatgen/issues/3684. Feel free to ping me if there is anything I could help.

janosh commented:
we used to install `openbabel` in CI with `conda` but it was removed in mid 2022 due to causing frequent issues. see https://github.com/materialsproject/pymatgen/commit/b13c65bd59b1f2b55b769c928a8ab8af7e539181. i'm open to giving it another try but i'd prefer not to install with `conda` and instead follow the [build from source instructions](https://open-babel.readthedocs.io/en/latest/Installation/install.html#basic-build-procedure). `conda` is slow and not currently used in pymatgen CI so would have to be installed first just to install a single dep

orionarcher commented:
I'll try building from source.

We could conda install openbabel, packmol, and the three required packages from the openff toolkit, which would be a little more worthwhile. A lower overhead option could be [micromamba](https://github.com/marketplace/actions/setup-micromamba) but I think that would require larger changes to the build script. Just a thought!

orionarcher commented:
Installing `openff-toolkit` from source requires separately installing all dependencies, some of which are only available on conda. I don't think there is a clean way to make testing work in CI without using either `conda` or `micromamba`.

janosh commented:
let's give `micromamba` a try then? interested to see how it performs here given it's meant to be faster than `conda`

orionarcher commented:
I switched over the test suite to use micromamba and it seems to be ~20-30% faster on ubuntu and comparable on windows.

This switch enabled me to install openbabel, which was not previously included in testing. I discovered several failing tests that were flying under the radar and that are now causing CI to fail.
- errors in `test_babel.py` which I fixed
- errors in `analysis/test_graphs.py` and `io/qchem/test_outputs.py` that I am not sure how to debug. These are failing both locally and in the ubuntu CI. It's not an issue in the windows CI because openbabel is not being installed.

@janosh, what do you recommend as the path forward here?

Should we:
a) merge as is with tests failing?
b) disable the tests and contact the maintainers?
c) have someone (preferably other than me) try to debug the issue in this PR?

janosh commented:
luckily fixing the 3 failing tests was fairly easy. fixing the third surfaced a 4 issue so there's still 1 failing test

> b) disable the tests and contact the maintainers?

i think we can skip `TestQCOutput.test_all` for now and ping @samblau and @espottesmith. if you guys have time to take a look, feel free to suggest a fix

janosh reviewed (approved):
thanks @orionarcher. excited about the new OpenFF integration!

and really appreciate that you reinstated the `openbabel` tests!
